### PR TITLE
[FeatureHighlight] Implementation of Typography Themer.

### PR DESCRIPTION
**Pull request by mohammadcazig**
Typography themer Implemented:
Pivotal: https://www.pivotaltracker.com/story/show/156171627
New Target,
Tests,
Example,
screen shot 2018-04-06 at 11 42 24 am

### CODE REVIEW

**Comment by jverkoey** [comment]
Tests are failing.

**Comment by jverkoey** [comment]
```
components/FeatureHighlight/tests/unit/FeatureHighlightTypographyThemerTests.m:18:9: fatal error: 'MDCFeatureHighlightTypographyThemer.h' file not found
#import "MDCFeatureHighlightTypographyThemer.h"
        ^
1 error generated.
```

File: MaterialComponents.podspec [hunk 1]
```diff
@@ -335,6 +335,14 @@ Pod::Spec.new do |mdc|
       extension.dependency "MaterialComponents/FeatureHighlight"
       extension.dependency "MaterialComponents/Themes"
     end
+    component.subspec "Typography" do |extension|
```
Comment (jverkoey) [inline MaterialComponents.podspec]: TypographyThemer

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h [hunk 2]
```diff
@@ -122,6 +122,21 @@ typedef void (^MDCFeatureHighlightCompletion)(BOOL accepted);
...
+ Defaults to nil.
+ */
+@property(nonatomic, strong, nullable) UIFont *titleFont;
+
+/**
+ Sets the font to be used for the body text.
+
+ Defaults to nil.
+ */
+@property(nonatomic, strong, nullable) UIFont *bodyFont;
```
Comment (jverkoey) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: Pull these new APIs out to a separate PR please so that if we have to revert this PR we don't lose all of the functionality.

**Review by jverkoey — DISMISSED** [review]
Typos + suggestion to split PR into two.

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h (hunk 2 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: Done. #3274

File: components/FeatureHighlight/BUILD [hunk 3]
```diff
@@ -105,6 +117,7 @@ mdc_objc_library(
         ":ColorThemer",
         ":FontThemer",
         ":FeatureHighlightAccessibilityMutator",
+        ":TypographyThemer",
```
Comment (romoore) [inline components/FeatureHighlight/BUILD:120]: Is this a typo? I don't see any changes to the accessibility mutator.

File: components/FeatureHighlight/src/FontThemer/MDCFeatureHighlightFontThemer.h [hunk 4]
```diff
@@ -21,6 +21,7 @@
 
 /**
  Used to apply a font scheme to theme to MDCFeatureHighlightView.
+ This Class will be deprcated soon please use MDCFeatureHighlightTypographyThemer instead.
```
Comment (romoore) [inline components/FeatureHighlight/src/FontThemer/MDCFeatureHighlightFontThemer.h]: Nit: Spelling, `s/deprcated/deprecated/`.  I don't think "Class" needs to be capitalized.

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h [hunk 5]
```diff
@@ -122,6 +122,21 @@ typedef void (^MDCFeatureHighlightCompletion)(BOOL accepted);
  */
 @property(nonatomic, strong, nullable) UIColor *bodyColor;
 
+/**
+ Sets the font to be used for the title text.
```
Comment (romoore) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: Super-nit/suggestion: "The font for the title text."

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h [hunk 6]
```diff
@@ -122,6 +122,21 @@ typedef void (^MDCFeatureHighlightCompletion)(BOOL accepted);
...
+
+ Defaults to nil.
+ */
+@property(nonatomic, strong, nullable) UIFont *titleFont;
+
+/**
+ Sets the font to be used for the body text.
```
Comment (romoore) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: Super-nit/suggestion: "The font for the body text."

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h (hunk 2 above)
Comment (romoore) [reply components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: I still see these APIs in this PR. Is this a Work in Progress?

File: components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ Used to apply a typography scheme to theme to MDCFeatureHighlightView.
+ */
+@interface MDCFeatureHighlightTypographyThemer : NSObject
+
+/**
+ Applies a typography scheme to theme to a MDCFeatureHighlightView.
+ 
+ @param typographyScheme The typography scheme to apply to MDCFeatureHighlightView.
+ @param featureHighlightViewController A MDCFeatureHighlightViewController instance to apply a
```
Comment (romoore) [inline components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h]: Minor doc typo: Looks like you copied the "View" docs for the "VC" method?

File: components/FeatureHighlight/tests/unit/FeatureHighlightTypographyThemerTests.m [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+- (void)tearDown {
+  [super tearDown];
+  self.showView = nil;
+  self.highlightedView = nil;
+}
+
+- (void)testFeaturehighlightTypographyThemer {
+  MDCFeatureHighlightViewController *featureHighlightViewController =
+      [[MDCFeatureHighlightViewController alloc] initWithHighlightedView:self.highlightedView
```
Comment (romoore) [inline components/FeatureHighlight/tests/unit/FeatureHighlightTypographyThemerTests.m:44]: These arguments don't make sense.  Consider just passing `[[UIView alloc] init]` for each, or use `-initWithHighlightedView:completion:` instead.

* `highlightedView` is the "focus" of the Feature Highlight
* `showView` is used to have a "live" version of a view on top of the Feature Highlight. This is an alternative to a "screenshot" of `highlightView` when a live view is required. It's not supposed to be a subview of the `highlightedView`, but perhaps a "clone" instead.  See `FeatureHighlightShownViewExample` for an example usage.

**Review by romoore — DISMISSED** [review]

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.h (hunk 2 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: I am waiting for the other PR to get reviewed.

Comment (romoore) [reply components/FeatureHighlight/src/MDCFeatureHighlightViewController.h]: Sorry, I think I misread your comment:

> tests are fixed, separated PR has been submitted.

File: components/FeatureHighlight/src/MDCFeatureHighlightViewController.m [hunk 9]
```diff
@@ -192,6 +192,22 @@ - (void)setBodyColor:(UIColor *)bodyColor {
   self.view.bodyColor = bodyColor;
 }
 
+- (void)setTitleFont:(UIFont *)titleFont {
+  self.view.titleFont = titleFont;
```
Comment (ianegordon) [inline components/FeatureHighlight/src/MDCFeatureHighlightViewController.m]: This will call loadView, which is likely unexpected.
Please maintain local ivars for all of these properties and see if view exists before setting.
Additionally, set these view.properties in viewDidLoad.

**Review by ianegordon — DISMISSED** [review]

File: components/FeatureHighlight/BUILD (hunk 3 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/BUILD:120]: It's actually added to "unit_test_sources".

File: components/FeatureHighlight/src/FontThemer/MDCFeatureHighlightFontThemer.h (hunk 4 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/src/FontThemer/MDCFeatureHighlightFontThemer.h]: done.

File: components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h (hunk 7 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h]: done.

File: components/FeatureHighlight/tests/unit/FeatureHighlightTypographyThemerTests.m (hunk 8 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/tests/unit/FeatureHighlightTypographyThemerTests.m:44]: got rid of adding the subview, `init` is not allowed for this class and the `initWithHighlightedView:completion:` tries to take snapshot after update which crashes for this test, since there is no view on the display.

**Review by jverkoey — APPROVED** [review]

File: components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+#import "MaterialFeatureHighlight.h"
+#import "MaterialTypographyScheme.h"
+
+/**
+ Used to apply a typography scheme to theme to MDCFeatureHighlightViewController.
+ */
+@interface MDCFeatureHighlightTypographyThemer : NSObject
+
+/**
+ Applies a typography scheme to theme to a MDCFeatureHighlightViewController.
```
Comment (romoore) [inline components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h]: Super-nit: s/a MDCFeatureHighlightViewController/an MDCFeatureHighlightViewController/

Here and below

**Review by romoore — APPROVED** [review]

File: components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h (hunk 10 above)
Comment (mohammadcazig) [reply components/FeatureHighlight/src/TypographyThemer/MDCFeatureHighlightTypographyThemer.h]: done.
